starlark/hashtable.go

```diff
@@ -371,3 +364,186 @@ func softHashString(s string) uint32 {
 	}
 	return h
 }
+
+// An OrderedStringDict is a mapping from names to values with support
+// for fast indexing. Keys are stored in order of unique insertion.
+// Keys are fast to add and access but slow to delete.
+// It is not a true starlark.Value.
+type OrderedStringDict struct {
+	// Implementation based on the above hashtable.
+	table   []osdBucket     // len is zero or a power of two
+	bucket0 [bucketSize]int //*osdEntry // inline allocation for small maps.
+	entries []osdEntry      // sorted list of entries
+}
+
+type osdBucket []int //*osdEntry
```

Comment:
// each element is the index into entries


Use uint32 to save space (and panic if len >= 4G).